flink-staging/flink-ml/src/main/scala/org/apache/flink/ml/optimization/RegularizationType.scala

```diff
@@ -0,0 +1,116 @@
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.flink.ml.optimization
+
+import breeze.numerics._
+import org.apache.flink.ml.math.{BLAS, Vector}
+import org.apache.flink.ml.math.Breeze._
+import breeze.linalg.{norm => BreezeNorm, Vector => BreezeVector, max}
+
+// TODO(tvas): Change name to RegularizationPenalty?
+abstract class RegularizationType extends Serializable{
+  /** Calculates and applies the regularization amount and the regularization parameter
+    *
+    * @param weights The old weights
+    * @param effectiveStepSize The effective step size for this iteration
+    * @param regParameter The current regularization parameter
+    * @return A tuple whose first element is the updated weight vector and the second is the
+    *         new regularization parameter
+    */
+  def applyRegularization(weights: Vector, effectiveStepSize: Double, regParameter: Double):
+  (Vector, Double)
+  // TODO(tvas): We are not currently using the regularization value anywhere, but it could be
+  // useful to keep a history of it.
+
+}
+
+class NoRegularization extends RegularizationType {
+  /** Returns the original weights without any regularization applied
+    *
+    * @param weights The old weights
+    * @param effectiveStepSize The effective step size for this iteration
+    * @param regParameter The current regularization parameter
+    * @return A tuple whose first element is the updated weight vector and the second is the
+    *         regularization value
+    */
+  override def applyRegularization(weights: Vector,
+                                   effectiveStepSize: Double,
+                                   regParameter: Double):
+  (Vector, Double) = {(weights, 0.0)}
+}
+
+class L2Regularization extends RegularizationType {
+  /** Calculates and applies the regularization amount and the regularization parameter
+    *
+    * Implementation was taken from the Apache Spark Mllib library:
+    * http://git.io/vfZIT
+    * @param weights The old weights
+    * @param effectiveStepSize The effective step size for this iteration
+    * @param regParameter The current regularization parameter
+    * @return A tuple whose first element is the updated weight vector and the second is the
+    *         regularization value
+    */
+  override def applyRegularization(weights: Vector,
+                                   effectiveStepSize: Double,
+                                   regParameter: Double):
+  (Vector, Double) = {
+
+    val brzWeights: BreezeVector[Double] = weights.asBreeze
+    brzWeights :*= (1.0 - effectiveStepSize * regParameter)
```

Comment:
I think you have to apply the regularization gradient on the old weight vector and not on the new weight vector to make it mathematically sound here.